go/vt/sidecardb/sidecardb.go

```diff
@@ -0,0 +1,373 @@
+		_, err = si.exec(si.ctx, fmt.Sprintf("use %s", dbName), 1000, false, false)
+		if err != nil {
+			return "", err
+		}
+	}
+	return currentDB, nil
+}
+
+// gets existing schema of table
+func (si *vtSchemaInit) getCurrentSchema(tableName string) (string, error) {
+	var currentTableSchema string
+
+	rs, err := si.exec(si.ctx, fmt.Sprintf(ShowCreateTableQuery, tableName), 1, false, false)
+	if err != nil {
+		log.Errorf("Error getting _vt table schema for %s: %+v", tableName, err)
+		return "", err
+	}
+	if len(rs.Rows) > 0 {
+		currentTableSchema = rs.Rows[0][1].ToString()
+	}
+	return currentTableSchema, nil
+}
+
+// remove the extra DEFAULT CHARSET which we get from "show create table"
+func stripCharset(schema string) string {
+	ind := strings.Index(schema, "DEFAULT CHARSET")
+	if ind <= 0 {
+		return schema
+	}
+	return schema[:ind]
+}
+
+// finds diff that needs to be applied to current table schema to get the desired one. Will be an empty string if they match.
+// could be a create statement if the table does not exist or an alter if table exists but has a different schema
+func (si *vtSchemaInit) findTableSchemaDiff(current, desired string) (string, error) {
+	// todo: can we fix schemadiff to avoid the need for this?
+	current = stripCharset(current)
+
+	hints := &schemadiff.DiffHints{}
+	diff, err := schemadiff.DiffCreateTablesQueries(current, desired, hints)
+	if err != nil {
+		return "", err
+	}
+
+	tableAlterSQL := diff.CanonicalStatementString()
+	if strings.Contains(tableAlterSQL, "ALTER") {
+		// temp logging, should be removed in v17
+		log.Infof("alter sql %s", tableAlterSQL)
+		log.Infof("current schema %s", current)
+	}
+	return tableAlterSQL, nil
+}
+
+// createOrUpgradeTable expects that we are already in the _vt database
+func (si *vtSchemaInit) createOrUpgradeTable(table *vtTable) error {
+	var desiredTableSchema string
+	ctx := si.ctx
+	bytes, err := schemaLocation.ReadFile(table.path)
+	if err != nil {
+		return err
+	}
+	desiredTableSchema = string(bytes)
+
+	var tableAlterSQL string
+	tableExists := si.tableExists(table.name)
+	if tableExists {
+		currentTableSchema, err := si.getCurrentSchema(table.name)
+		if err != nil {
+			return err
+		}
+		tableAlterSQL, err = si.findTableSchemaDiff(currentTableSchema, desiredTableSchema)
+		if err != nil {
+			return err
+		}
+	} else {
+		tableAlterSQL = desiredTableSchema
+	}
+
+	if strings.TrimSpace(tableAlterSQL) != "" {
+		if !si.dbCreated {
+			// We use CreateVTDatabaseQuery to also create the first binlog entry when a primary comes up.
+			// That statement doesn't make it to the replicas, so we run the query again so that it is replicated
+			// to the replicas so that the replicas can create the _vt database.
+			if err := si.createVTDatabase(); err != nil {
+				return err
+			}
+			si.dbCreated = true
+		}
+		_, err = si.exec(ctx, tableAlterSQL, 1, false, true)
+		if err != nil {
+			if strings.Contains(err.Error(), "already exists") {
```

Comment:
better to cast to `SqlErr` and get the actual error code (see `mysql/constants.go`)

Reply:
done